overlord/servicestate/quota_control.go

```diff
@@ -473,5 +486,48 @@ func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) erro
 	}
 
 	// ensure service states are updated
-	return ensureSnapServicesForGroup(st, grp, allGrps)
+	return ensureSnapServicesForGroup(st, grp, allGrps, nil)
+}
+
+// EnsureSnapAbsentFromQuotaGroup ensures that the specified snap is not present
+// in any quota group, usually in preparation for removing that snap from the
+// system to keep the quota group itself consistent.
+func EnsureSnapAbsentFromQuotaGroup(st *state.State, snap string) error {
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	// try to find the snap in any group
+	var grp *quota.Group
+	for _, maybeGrp := range allGrps {
+		if strutil.ListContains(maybeGrp.Snaps, snap) {
+			grp = maybeGrp
```

Comment:
Can we just iterate over maybeGrp.Snaps "manually" (not via ListContains) keeping the iterator and on matching snap just drop it by something like this:
```
grp.Snaps[i] = grp.Snaps[len(grp.Snaps)-1]
grp.Snaps = grp.Snaps[:len(grp.Snaps)-1]
```
(replaces matching snap with last snap on the list, then drops the last one).

This will be much shorter and easy to read, and avoids newSnapList below.

Reply:
I guess I prefer my implementation since (to me) it is simpler because it doesn't involve index calculations. It is certainly less efficient, but I think it's less prone to off by one errors since the only index calculation we do is to subtract one from the length when initializing the capacity of the new list. For comparison, these are the two implementations:

```go
	var updatedGroup *quota.Group
	found := false
	for _, grp := range allGrps {
		for idx, sn := range grp.Snaps {
			if sn == snap {
				// drop this snap from the list of Snaps by swapping it with the
				// last snap in the list, and then dropping the last snap from
				// the list
				grp.Snaps[idx] = grp.Snaps[len(grp.Snaps)-1]
				grp.Snaps = grp.Snaps[:len(grp.Snaps)-1]
				updatedGroup = grp
				found = true
				break
			}
		}
		if found {
			break
		}
	}

	if !found {
		return nil
	}
```

compared to my current implementation:
```go
	var grp *quota.Group
	for _, maybeGrp := range allGrps {
		if strutil.ListContains(maybeGrp.Snaps, snap) {
			grp = maybeGrp
			break
		}
	}

	if grp == nil {
		return nil
	}

	// drop this snap from the snaps list of the group
	newSnapList := make([]string, 0, len(grp.Snaps)-1)
	for _, sn := range grp.Snaps {
		if sn != snap {
			newSnapList = append(newSnapList, sn)
		}
	}

	grp.Snaps = newSnapList
```

IMHO the latter is easier to understand and doesn't involve any math to keep track of any index calculations and is also less deeply nested

Reply:
Fair enough, I think I slightly disagree ;) but it's definitely a bikeshedding material and not a blocker, the code is of course correct. FWTW the two variants you gave could still be simplified by:

- breaking from nested loop with labeled break and eliminating `if found.. check`.
- not having 'found' variable because updatedGrp != nil means we found something
- or even actually moving patchQuotas and ensureSnapServicesForGroup to the loop because we know we need one match only, so as soon as we find it we can update the group and apply the changes, then return from the same spot.

Reply:
I hadn't thought about putting the patchQuotas and ensureSnapServicesForGroup inside the loop like this, I think this makes it simpler than my solution even though it does still have some clever index calculations

Reply:
I like the new implementation. I am (slightly) mad at go for not having "delete(l slice, index int)" but yeah, one gets used to this and it's okay to read IMHO.

Reply:
yes I cannot count how many times I get annoyed that there is not a `delete`